ipa-core/src/protocol/ipa_prf/mod.rs

```diff
@@ -192,7 +192,10 @@ where
     Replicated<TS>: BooleanArrayMul,
     Replicated<TV>: BooleanArrayMul,
     F: PrimeField + ExtendableField,
-    Replicated<F>: Serializable,
+    Replicated<F>: Serializable + SecureMul<<C as UpgradableContext>::UpgradedContext<F>>,
+    Replicated<Boolean>: SecureMul<C>,
+    Replicated<Fp25519, 64>: SecureMul<C>,
```

Comment:
I agree that the trait bounds need to be fixed. I would like to just add the trait bound C:Upgradable. I would then implement SecureMul for all Boolean types for <C as Upgradable>::DZKPUpgraded. Further, implement SecureMul for all primefield types for <C as Upgradable>::Upgraded. This would require to change all functions to use the new maliciousAPI and upgrade the context whenever multiplications are used. I believe this would resolve all the trait bounds but I am not quite there yet.

Reply:
https://github.com/private-attribution/ipa/issues/1022